incubator/hnc/internal/validators/namespace_test.go

```diff
@@ -12,29 +12,53 @@ import (
 )
 
 func TestDeleteSubNamespace(t *testing.T) {
-	// Create a namespace with owner annotation.
+	// Create a namespace instance a.
 	sub := &corev1.Namespace{}
 	sub.Name = "a"
-	setSubAnnotation(sub, "b")
 
-	// a (subnamespace of b) -> b
-	f := foresttest.Create("B-")
-	vns := &Namespace{Forest: f}
+	// We will put an anchor a in the namespace c if it exists.
+	tests := []struct {
+		name           string
+		forest         string
+		subnamespaceOf string
+		fail           bool
+	}{
+		// There's no test case for when parent and anchor don't match, because the
+		// reconciler always sets the parent to match the subnamespaceOf annotation.
+		// - a (subnamespaceOf b)
+		{name: "when parent is missing", forest: "-", subnamespaceOf: "b"},
+		// - a (subnamespaceOf b) -> b (no anchor)
+		{name: "when anchor is missing", forest: "b-", subnamespaceOf: "b"},
+		// - a (subnamespaceOf c) -> c (has anchor a), b
+		{name: "when annotation and anchor match", forest: "c--", subnamespaceOf: "c", fail: true},
+		// - a (subnamespaceOf c) -> b, c(has anchor a)
+		{name: "when annotation and anchor don't match", forest: "b--", subnamespaceOf: "b"},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			g := NewGomegaWithT(t)
+			setSubAnnotation(sub, tc.subnamespaceOf)
+			req := &nsRequest{
+				ns: sub,
+				op: v1beta1.Delete,
+			}
 
-	t.Run("Delete namespace with owner annotation", func(t *testing.T) {
-		g := NewGomegaWithT(t)
-		req := &nsRequest{
-			ns: sub,
-			op: v1beta1.Delete,
-		}
+			// Construct the forest
+			f := foresttest.Create(tc.forest)
+			// Add anchor a to namespace c if it exists.
```

Comment:
This needs more comments somewhere, either here or maybe on lines 32 and 34. I can see that we _are_ adding the anchor if `c` exists, but _why_ are we doing this? (I figured it out, I'd just like it to be a bit better documented in the code).

Reply:
I added the reason here in the same line.